pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java

```diff
@@ -0,0 +1,320 @@
+        }
+        break;
+      case BYTES:
+        if (_rightDataType != FieldSpec.DataType.BYTES) {
+          throw new IllegalStateException(String.format(
+              "Unsupported data type for comparison: [Left Transform Function [%s] result type is [%s], Right Transform Function [%s] result type is [%s]]",
+              _leftTransformFunction.getName(), _leftDataType, _rightTransformFunction.getName(), _rightDataType));
+        }
+        break;
+      default:
+        throw new IllegalStateException(String.format(
+            "Unsupported data type for comparison: [Left Transform Function [%s] result type is [%s], Right Transform Function [%s] result type is [%s]]",
+            _leftTransformFunction.getName(), _leftDataType, _rightTransformFunction.getName(), _rightDataType));
+    }
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return INT_SV_NO_DICTIONARY_METADATA;
+  }
+
+  protected void fillResultArray(ProjectionBlock projectionBlock) {
+    if (_results == null) {
+      _results = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    int length = projectionBlock.getNumDocs();
+    switch (_leftDataType) {
+      case INT:
+        int[] leftIntValues = _leftTransformFunction.transformToIntValuesSV(projectionBlock);
+        switch (_rightDataType) {
+          case INT:
+            int[] rightIntValues = _rightTransformFunction.transformToIntValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(Integer.compare(leftIntValues[i], rightIntValues[i]));
+            }
+            break;
+          case LONG:
+            long[] rightLongValues = _rightTransformFunction.transformToLongValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(Long.compare(leftIntValues[i], rightLongValues[i]));
+            }
+            break;
+          case FLOAT:
+            float[] rightFloatValues = _rightTransformFunction.transformToFloatValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(Double.compare(leftIntValues[i], rightFloatValues[i]));
+            }
+            break;
+          case DOUBLE:
+            double[] rightDoubleValues = _rightTransformFunction.transformToDoubleValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(Double.compare(leftIntValues[i], rightDoubleValues[i]));
+            }
+            break;
+          case STRING:
+            String[] rightStringValues = _rightTransformFunction.transformToStringValuesSV(projectionBlock);
+            for (int i = 0; i < length; i++) {
+              _results[i] = getBinaryFuncResult(
+                  BigDecimal.valueOf(leftIntValues[i]).compareTo(new BigDecimal(rightStringValues[i])));
```

Comment:
Is the Transform function expected to treat operands as numbers and not comparable strings?

Reply:
Yes, if left side is int then right side should be a number, otherwise it's definitely wrong.
We only compare strings when both left and right sides are String.